add some storage layer codes to the switch APIs

soramichi opened the pull request:
Newly supported features: set/get default vlan id, set/get source mac address

Example API calls:
  sai_vlan_port_t port0 = { 0, SAI_VLAN_PORT_TAGGED };
  sai_vlan_port_t port1 = { 1, SAI_VLAN_PORT_TAGGED };
  sai_vlan_port_t ports[] = { port0, port1 };

  switch_api->initialize_switch(...);
  // at this point, all ports are in vlan 1 (default vlan)

  vlan_api->create_vlan(10);
  vlan_api->add_ports_to_vlan(10, 2, ports);
  // at this point, port0 and port1 are in vlan 10, whereas the other ports are still in vlan 1

  sai_attribute_t attr = {
    .id = SAI_SWITCH_ATTR_DEFAULT_PORT_VLAN_ID,
    .value.u16 = 2
  };
  switch_api->set_switch_attribute(&attr);
  // at this poitn, all ports except port0 and port1 are in vlan 2 (new default vlan)

Example Output:
  [SAI_SWITCH.NOTICE] stub_sai_switch.c[487]- stub_initialize_switch: Initialize switch
  [SAI_VLAN.NOTICE] stub_sai_vlan.c[185]- stub_create_vlan: Create vlan vlan 1
  [SAI_VLAN.NOTICE] stub_sai_vlan.c[339]- stub_add_ports_to_vlan: Ports added to VLAN(1): 0 1 2 3 4 5 6 7 8 9
  [SAI_VLAN.NOTICE] stub_sai_vlan.c[345]- stub_add_ports_to_vlan: Vlan assigned to each port: 1 1 1 1 1 1 1 1 1 1
  [SAI_VLAN.NOTICE] stub_sai_vlan.c[185]- stub_create_vlan: Create vlan vlan 10
  [SAI_VLAN.NOTICE] stub_sai_vlan.c[339]- stub_add_ports_to_vlan: Ports added to VLAN(10): 0 1
  [SAI_VLAN.NOTICE] stub_sai_vlan.c[345]- stub_add_ports_to_vlan: Vlan assigned to each port: 10 10 1 1 1 1 1 1 1 1
  [SAI_UTILS.NOTICE] stub_sai_utils.c[244]- set_dispatch_attrib_handler: Set Switch port default vlan ID, key:, val:2
  [SAI_VLAN.NOTICE] stub_sai_vlan.c[185]- stub_create_vlan: Create vlan vlan 2
  [SAI_SWITCH.NOTICE] stub_sai_switch.c[659]- stub_switch_default_port_vlan_set: Default vlan changed from 1 to 2
  [SAI_VLAN.NOTICE] stub_sai_vlan.c[339]- stub_add_ports_to_vlan: Ports added to VLAN(2): 2 3 4 5 6 7 8 9
  [SAI_VLAN.NOTICE] stub_sai_vlan.c[345]- stub_add_ports_to_vlan: Vlan assigned to each port: 10 10 2 2 2 2 2 2 2 2
  [SAI_VLAN.NOTICE] stub_sai_vlan.c[268]- stub_remove_vlan: Remove vlan vlan 1

itaibaz commented:
Hi,
One comment re stub_switch_port_list_get

In SAI list usage in get :
* - In get attribute function call, the function call returns a list of objects
* to the caller in the list member. The caller is responsible for allocating the
* buffer for the list member and set the count member to the size of allocated object
* list. If the size is large enough to accomodate the list of object id, the
* callee will then fill the list member and set the count member to the actual
* number of objects.  If the list size is not large enough, the callee will set the
* count member to the actual number of object id and return
* SAI_STATUS_BUFFER_OVERFLOW. Once the caller gets such return code, it should
* use the returned count member to re-allocate list and retry.
*/

While you allocated the list inside stub_switch_port_list_get
I suggest using stub_fill_objlist utility function, you can find example in stub_next_hop_group_hop_list_get

Thanks

From: Soramichi Akiyama [mailto:notifications@github.com]
Sent: Tuesday, June 30, 2015 8:38 AM
To: opencomputeproject/OCP-Networking-Project-Community-Contributions
Subject: [OCP-Networking-Project-Community-Contributions] add some storage layer codes to the switch APIs (#117)


Newly supported features: set/get default vlan id, set/get source mac address

Example API calls:
sai_vlan_port_t port0 = { 0, SAI_VLAN_PORT_TAGGED };
sai_vlan_port_t port1 = { 1, SAI_VLAN_PORT_TAGGED };
sai_vlan_port_t ports[] = { port0, port1 };

switch_api->initialize_switch(...);
// at this point, all ports are in vlan 1 (default vlan)

vlan_api->create_vlan(10);
vlan_api->add_ports_to_vlan(10, 2, ports);
// at this point, port0 and port1 are in vlan 10, whereas the other ports are still in vlan 1

sai_attribute_t attr = {
.id = SAI_SWITCH_ATTR_DEFAULT_PORT_VLAN_ID,
.value.u16 = 2
};
switch_api->set_switch_attribute(&attr);
// at this poitn, all ports except port0 and port1 are in vlan 2 (new default vlan)

Example Output:
[SAI_SWITCH.NOTICE] stub_sai_switch.c[487]- stub_initialize_switch: Initialize switch
[SAI_VLAN.NOTICE] stub_sai_vlan.c[185]- stub_create_vlan: Create vlan vlan 1
[SAI_VLAN.NOTICE] stub_sai_vlan.c[339]- stub_add_ports_to_vlan: Ports added to VLAN(1): 0 1 2 3 4 5 6 7 8 9
[SAI_VLAN.NOTICE] stub_sai_vlan.c[345]- stub_add_ports_to_vlan: Vlan assigned to each port: 1 1 1 1 1 1 1 1 1 1
[SAI_VLAN.NOTICE] stub_sai_vlan.c[185]- stub_create_vlan: Create vlan vlan 10
[SAI_VLAN.NOTICE] stub_sai_vlan.c[339]- stub_add_ports_to_vlan: Ports added to VLAN(10): 0 1
[SAI_VLAN.NOTICE] stub_sai_vlan.c[345]- stub_add_ports_to_vlan: Vlan assigned to each port: 10 10 1 1 1 1 1 1 1 1
[SAI_UTILS.NOTICE] stub_sai_utils.c[244]- set_dispatch_attrib_handler: Set Switch port default vlan ID, key:, val:2
[SAI_VLAN.NOTICE] stub_sai_vlan.c[185]- stub_create_vlan: Create vlan vlan 2
[SAI_SWITCH.NOTICE] stub_sai_switch.c[659]- stub_switch_default_port_vlan_set: Default vlan changed from 1 to 2
[SAI_VLAN.NOTICE] stub_sai_vlan.c[339]- stub_add_ports_to_vlan: Ports added to VLAN(2): 2 3 4 5 6 7 8 9
[SAI_VLAN.NOTICE] stub_sai_vlan.c[345]- stub_add_ports_to_vlan: Vlan assigned to each port: 10 10 2 2 2 2 2 2 2 2
[SAI_VLAN.NOTICE] stub_sai_vlan.c[268]- stub_remove_vlan: Remove vlan vlan 1

________________________________
You can view, comment on, or merge this pull request online at:

  https://github.com/opencomputeproject/OCP-Networking-Project-Community-Contributions/pull/117

Commit Summary

  *   add some storage layer codes to the switch APIs

File Changes

  *   M sai/stub/inc/stub_sai.h<https://github.com/opencomputeproject/OCP-Networking-Project-Community-Contributions/pull/117/files#diff-0> (31)
  *   M sai/stub/src/stub_sai_switch.c<https://github.com/opencomputeproject/OCP-Networking-Project-Community-Contributions/pull/117/files#diff-1> (101)
  *   M sai/stub/src/stub_sai_vlan.c<https://github.com/opencomputeproject/OCP-Networking-Project-Community-Contributions/pull/117/files#diff-2> (64)

Patch Links:

  *   https://github.com/opencomputeproject/OCP-Networking-Project-Community-Contributions/pull/117.patch
  *   https://github.com/opencomputeproject/OCP-Networking-Project-Community-Contributions/pull/117.diff

—
Reply to this email directly or view it on GitHub<https://github.com/opencomputeproject/OCP-Networking-Project-Community-Contributions/pull/117>.

soramichi commented:
Hi,

thanks for the comments.
I fixed the codes to let the user allocate memory for retrieving the port list.
(also removed unused struct made by mistake.)

soramichi commented:
Hi, is it possible to get some feedback (should be OK, should be fixed more, or whatever) on this PR?

itaibaz commented:
Hi,
In port list get, there are few things that can be improved :

Re

  if ((int)value->objlist.count < number_of_ports) {



+        STUB_LOG_NTC("The given buffer does not have sufficient length\n");


+      value->objlist.count = number_of_ports;


+      return SAI_STATUS_BUFFER_OVERFLOW;


+    }


This logic is redundant, as it is done by stub_fill_objlist

Re
+    else if (value->objlist.list == NULL) {



+        STUB_LOG_ERR("The given buffer points an invalid address\n");


+      return SAI_STATUS_INVALID_PARAMETER;


+    }

This logic is redundant, as it is done by check_attribs_metadata
Which always checks list is not null, for all types holding a list

+    value->objlist.count = number_of_ports;
The count is also filled by stub_fill_objlist

So, in the end, the only code needed is
+    ret = stub_fill_objlist(ports, number_of_ports, &value->objlist);

The intention of the utils infra is to implement all the common logic, and really minimize the amount of code you need to write for each getter/setter.

Thanks


From: Soramichi Akiyama [mailto:notifications@github.com]
Sent: Tuesday, July 07, 2015 7:15 AM
To: opencomputeproject/OCP-Networking-Project-Community-Contributions
Cc: Itai Baz
Subject: Re: [OCP-Networking-Project-Community-Contributions] add some storage layer codes to the switch APIs (#117)


Hi, is it possible to get some feedback (should be OK, should be fixed more, or whatever) on this PR?

—
Reply to this email directly or view it on GitHub<https://github.com/opencomputeproject/OCP-Networking-Project-Community-Contributions/pull/117#issuecomment-119068417>.

soramichi commented:
Thank you for the review and sorry for not checking the existing implementation carefully.
I removed the redundant codes.

itaibaz commented:
Hi,
Thanks for the fix, I think the code is ready to be merged.
Guohan, can you please review and merge

Thanks


From: Soramichi Akiyama [mailto:notifications@github.com]
Sent: Wednesday, July 08, 2015 4:37 AM
To: opencomputeproject/OCP-Networking-Project-Community-Contributions
Cc: Itai Baz
Subject: Re: [OCP-Networking-Project-Community-Contributions] add some storage layer codes to the switch APIs (#117)


Thank you for the review and sorry for not checking the existing implementation carefully.
I removed the redundant codes.

—
Reply to this email directly or view it on GitHub<https://github.com/opencomputeproject/OCP-Networking-Project-Community-Contributions/pull/117#issuecomment-119395366>.

lguohan commented:
vlan_api->add_ports_to_vlan(10, 2, ports);
// at this point, port0 and port1 are in vlan 10, whereas the other ports are still in vlan 1

In this case, port0 and port1 should be in both vlan 10 and vlan 1, where vlan10 is tagged mode and vlan 1 is untagged mode.

soramichi commented:
Thank you for pointing out the mistake.
I fixed the codes to deal correctly with the tagged/untagged mode ports.

New outputs for the same API calls are as follows ('u' means untagged and 't' means tagged).
When port0 and port1 are added to VLAN10 with tagged mode, they remain also in VLAN1 with untagged mode.
Also, when the default vlan id is changed from 1 to 2, VLAN10 keeps including port0 and port1 with tagged mode.

    [SAI_SWITCH.NOTICE] stub_sai_switch.c[487]- stub_initialize_switch: Initialize switch
    [SAI_VLAN.NOTICE] stub_sai_vlan.c[185]- stub_create_vlan: Create vlan vlan 1
    [SAI_VLAN.NOTICE] stub_sai_vlan.c[349]- stub_add_ports_to_vlan: Ports in VLAN(1): 0(u) 1(u) 2(u) 3(u) 4(u) 5(u) 6(u) 7(u) 8(u) 9(u)
    [SAI_VLAN.NOTICE] stub_sai_vlan.c[185]- stub_create_vlan: Create vlan vlan 10
    [SAI_VLAN.NOTICE] stub_sai_vlan.c[349]- stub_add_ports_to_vlan: Ports in VLAN(1): 0(u) 1(u) 2(u) 3(u) 4(u) 5(u) 6(u) 7(u) 8(u) 9(u)
    [SAI_VLAN.NOTICE] stub_sai_vlan.c[349]- stub_add_ports_to_vlan: Ports in VLAN(10): 0(t) 1(t)
    [SAI_UTILS.NOTICE] stub_sai_utils.c[244]- set_dispatch_attrib_handler: Set Switch port default vlan ID, key:, val:2
    [SAI_VLAN.NOTICE] stub_sai_vlan.c[185]- stub_create_vlan: Create vlan vlan 2
    [SAI_SWITCH.NOTICE] stub_sai_switch.c[664]- stub_switch_default_port_vlan_set: Default vlan changed from 1 to 2
    [SAI_VLAN.NOTICE] stub_sai_vlan.c[349]- stub_add_ports_to_vlan: Ports in VLAN(1):
    [SAI_VLAN.NOTICE] stub_sai_vlan.c[349]- stub_add_ports_to_vlan: Ports in VLAN(10): 0(t) 1(t)
    [SAI_VLAN.NOTICE] stub_sai_vlan.c[349]- stub_add_ports_to_vlan: Ports in VLAN(2): 0(u) 1(u) 2(u) 3(u) 4(u) 5(u) 6(u) 7(u) 8(u) 9(u)
    [SAI_VLAN.NOTICE] stub_sai_vlan.c[268]- stub_remove_vlan: Remove vlan vlan 1

lguohan commented on sai/stub/src/stub_sai_switch.c:
```diff
@@ -610,6 +647,32 @@ sai_status_t stub_switch_default_port_vlan_set(_In_ const sai_object_key_t
+    stub_switch->default_port_vlan_id = vlan_id_new;
+    stub_remove_vlan(vlan_current->id);
+
     STUB_LOG_EXIT();
```
we have removed the SWITCH_ATTR_DEFAULT_PORT_VLAN because we feel this brings complex when user tries to change this value. 

See the commit ids: 24dae13, 3a08b37.
